### PR TITLE
Add Support of Interpose Chain Element to Cross Connect Endpoints

### PR DESCRIPTION
**Pull request by haiodo**
Implementation for #369 adding support for cross connect NSEs.

### CODE REVIEW

File: pkg/networkservice/common/crossnse/server.go [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+				crossCTX := clienturl.WithClientURL(ctx, crossNSEURL)
+
+				// Store client connection and selected cross connection URL.
+				_, _ = l.activeConnection.LoadOrStore(conn.Id, &connectionInfo{
+					endpointURL: clientURL,
+					crossNSEURL: crossNSEURL,
+				})
+				result, err = next.Server(crossCTX).Request(crossCTX, request)
+				if err != nil {
+					trace.Log(ctx).Errorf("failed to request cross NSE %v err: %v", crossNSEURL, err)
```
Comment (denis-tingajkin) [inline pkg/networkservice/common/crossnse/server.go:111]: Do we need to call `next.Server(crossCTX).Close()`?

Comment (haiodo) [reply pkg/networkservice/common/crossnse/server.go:111]: Yes, we need, Thanks.

Comment (haiodo) [reply pkg/networkservice/common/crossnse/server.go:111]: Nope, it is not required.

File: pkg/networkservice/common/crossnse/server.go [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package crossnse provides a NetworkServiceServer chain element that tracks local Cross connect Endpoints and call them first
+// their unix file socket as the clienturl.ClientURL(ctx) used to connect to them.
+package crossnse
```
Comment (edwarnicke) [inline pkg/networkservice/common/crossnse/server.go:19]: Hmm... I wonder if we can come up with a better name here for the package.  Functionally... what we are doing is interposing xconn NSEs... 

perhaps something like 

'interpose'

if generic?  Totally open to discussion here.  Good naming will make things much easier for subsequent consumers.

File: pkg/networkservice/common/crossnse/server.go [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+	"github.com/networkservicemesh/sdk/pkg/networkservice/core/next"
+)
+
+type crossConnectNSEServer struct {
+	// Map of names -> *registry.NetworkServiceEndpoint for local bypass to file crossNSEs
+	crossNSEs sync.Map
+
+	activeConnection sync.Map
+
+	nsmgrName string
```
Comment (edwarnicke) [inline pkg/networkservice/common/crossnse/server.go:46]: Question... why do these have nsmgr in their names?

File: pkg/networkservice/common/crossnse/server.go [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+	conn := request.GetConnection()
+	ind := conn.GetPath().GetIndex() // It is designed to be used inside Endpoint, so current index is Endpoint already
+	connID := conn.GetId()
+
+	if len(conn.GetPath().GetPathSegments()) == 0 || ind <= 0 {
+		return nil, errors.Errorf("path segment doesn't have a client or cross connect nse identity")
+	}
+	// Remember current segment Name if not yet.
+	l.nsmgrInit.Do(func() {
+		l.nsmgrName = conn.GetPath().GetPathSegments()[ind].Name
```
Comment (edwarnicke) [inline pkg/networkservice/common/crossnse/server.go:82]: I presume this is intended to happen after updatepath and so you are piggy backing on update path knowing the name?

File: pkg/networkservice/common/crossnse/server.go [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+	"github.com/networkservicemesh/api/pkg/api/networkservice"
+	"github.com/networkservicemesh/api/pkg/api/registry"
+
+	"github.com/networkservicemesh/sdk/pkg/networkservice/common/clienturl"
+	"github.com/networkservicemesh/sdk/pkg/networkservice/core/next"
+)
+
+type crossConnectNSEServer struct {
+	// Map of names -> *registry.NetworkServiceEndpoint for local bypass to file crossNSEs
+	crossNSEs sync.Map
```
Comment (edwarnicke) [inline pkg/networkservice/common/crossnse/server.go:42]: Was there a particular reason not to generate typed sync.Maps similar to https://github.com/networkservicemesh/sdk/blob/master/pkg/tools/clientmap/gen.go ?

File: pkg/networkservice/common/crossnse/server.go (hunk 2 above)
Comment (haiodo) [reply pkg/networkservice/common/crossnse/server.go:19]: interpose looks good for me :-)

File: pkg/networkservice/common/crossnse/server.go (hunk 3 above)
Comment (haiodo) [reply pkg/networkservice/common/crossnse/server.go:46]: Fixed.

File: pkg/networkservice/common/crossnse/server.go (hunk 4 above)
Comment (haiodo) [reply pkg/networkservice/common/crossnse/server.go:82]: Fixed, passed name to chain element, to simplify.

File: pkg/networkservice/common/connect/DESIGN.md [hunk 6]
```diff
@@ -27,7 +27,7 @@ calls a 'cancel' function provided to it at its construction.
 
 ## connectServer
 
-connectServer keeps clientsByID [clientmap.Map](https://github.com/networkservicemesh/sdk/blob/master/pkg/tools/clientmap/gen.go#L27) mapping incoming server Connection.ID
```
Comment (denis-tingajkin) [inline pkg/networkservice/common/connect/DESIGN.md:30]: Do we need this change in `DESIGN.md`?

File: pkg/registry/common/interpose/server.go [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+	"github.com/google/uuid"
+
+	"github.com/networkservicemesh/sdk/pkg/registry/core/next"
+
+	"github.com/golang/protobuf/ptypes/empty"
+	"github.com/networkservicemesh/api/pkg/api/registry"
+)
+
+// InterposeNSEName - a common prefix for all registered cross NSEs
+const InterposeNSEName = "cross-connect-nse#"
```
Comment (denis-tingajkin) [inline pkg/registry/common/interpose/server.go:38]: I think we can make this thing as a private const. For this, we need to add a registry/interpose client chain element that just checks incoming NSE name and if it has not this pattern then it should just append it. What do you think?

File: pkg/networkservice/common/connect/DESIGN.md (hunk 6 above)
Comment (haiodo) [reply pkg/networkservice/common/connect/DESIGN.md:30]: This PR is not changing connect server or any logic. So why?

File: pkg/registry/common/interpose/server.go (hunk 7 above)
Comment (haiodo) [reply pkg/registry/common/interpose/server.go:38]: This is a good Idea, I'd like it. Let me try. I think it should not check for name, but just replace it with Interpose required name.

File: pkg/registry/common/interpose/client.go [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+
+	"github.com/golang/protobuf/ptypes/empty"
+	"github.com/networkservicemesh/api/pkg/api/registry"
+)
+
+type interposeClient struct {
+}
+
+func (u *interposeClient) Register(ctx context.Context, in *registry.NetworkServiceEndpoint, opts ...grpc.CallOption) (*registry.NetworkServiceEndpoint, error) {
+	in.Name = interposeNSEName
```
Comment (denis-tingajkin) [inline pkg/registry/common/interpose/client.go:36]:
```suggestion
        if !stings.HasSuffix(in.Name, interposeNSEName) {
        	in.Name = interposeNSEName + in.Name 
        }
```

File: pkg/registry/common/interpose/client.go [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+	return next.NetworkServiceEndpointRegistryClient(ctx).Register(ctx, in, opts...)
+}
+
+func (u *interposeClient) Find(ctx context.Context, in *registry.NetworkServiceEndpointQuery, opts ...grpc.CallOption) (registry.NetworkServiceEndpointRegistry_FindClient, error) {
+	return next.NetworkServiceEndpointRegistryClient(ctx).Find(ctx, in, opts...)
+}
+
+func (u *interposeClient) Unregister(ctx context.Context, in *registry.NetworkServiceEndpoint, opts ...grpc.CallOption) (*empty.Empty, error) {
+	return next.NetworkServiceEndpointRegistryClient(ctx).Unregister(ctx, in, opts...)
```
Comment (denis-tingajkin) [inline pkg/registry/common/interpose/client.go:45]: Also I think we need to do this for unregister cases
```suggestion
        if !stings.HasSuffix(in.Name, interposeNSEName) {
        	in.Name = interposeNSEName + in.Name 
        }
	return next.NetworkServiceEndpointRegistryClient(ctx).Unregister(ctx, in, opts...)
```

File: pkg/networkservice/common/connect/DESIGN.md (hunk 6 above)
Comment (haiodo) [reply pkg/networkservice/common/connect/DESIGN.md]: Fixed.

File: pkg/registry/common/interpose/server.go (hunk 7 above)
Comment (haiodo) [reply pkg/registry/common/interpose/server.go]: Fixed.

File: pkg/networkservice/common/interpose/server.go [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+		})
+		if result != nil {
+			return result, nil
+		}
+		return nil, errors.Errorf("all cross NSE failed to connect to endpoint %v connection: %v", clientURL, conn)
+	}
+
+	// Go to endpoint URL
+	connInfo := connInfoRaw.(*connectionInfo)
+	crossCTX := clienturl.WithClientURL(ctx, connInfo.endpointURL)
```
Comment (Bolodya1997) [inline pkg/networkservice/common/interpose/server.go:121]: We already have client URL set either from discover or from localbypass chain element, is there any need to set it again?

Comment (haiodo) [reply pkg/networkservice/common/interpose/server.go:121]: Make seance, will fix.

File: pkg/registry/common/interpose/client.go (hunk 8 above)
Comment (haiodo) [reply pkg/registry/common/interpose/client.go:36]: Done.

File: pkg/registry/common/interpose/client.go (hunk 9 above)
Comment (haiodo) [reply pkg/registry/common/interpose/client.go:45]: Fixed.

File: pkg/registry/common/interpose/client.go (hunk 8 above)
Comment (denis-tingajkin) [reply pkg/registry/common/interpose/client.go:36]: The problem still actual.

Comment (haiodo) [reply pkg/registry/common/interpose/client.go]: Fixed.

File: pkg/networkservice/common/interpose/server.go (hunk 10 above)
Comment (haiodo) [reply pkg/networkservice/common/interpose/server.go:125]: Actually It would be better to check and verify if same endpointUrl is selected on both enters.

**Comment by denis-tingajkin** [comment]
@edwarnicke Do you have any additional comments for this PR?

**Pull request merged**